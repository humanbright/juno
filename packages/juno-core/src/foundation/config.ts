/* eslint-disable import/no-mutable-exports */
import { ShowDeprecatedArgs } from './utils';

/* eslint-disable no-console */
const JUNO_WARNING_IGNORE_KEY = '__JUNO_WARNING_IGNORE__';

type WarningFunctionType =
  | ((
      /** warning message */
      message: string,
      /** addition options with that warning */
      options?: Pick<ShowDeprecatedArgs, 'level'>,
    ) => void)
  // * make that type be optional to avoid production not have that method error
  | undefined;

interface RcConfigurationType {
  /** is ignore all warning show from juno */
  WARNING_IGNORE: boolean;
  WARNING_FUNCTION: WarningFunctionType;
}

// * use that way to avoid that code be exist in production code base
/**
 * config global setting
 *
 * ```ts
 * configure({
 *   warning: false
 * })
 * ```
 * ## WARNING related method and  only work in `development mode`
 *
 * `WARNING_IGNORE`:  only work in `development mode`
 * `WARNING_FUNCTION` only work in `development mode`
 */
let rcConfiguration = {} as RcConfigurationType;

if (process.env.NODE_ENV !== 'production') {
  class RcConfig implements RcConfigurationType {
    private _WARNING_IGNORE: boolean = false;

    private _WARNING_FUNCTION: WarningFunctionType = (
      message: string,
      options,
    ) => {
      console[options?.level || 'warn'](message);
    };

    constructor() {
      if (typeof window !== 'undefined') {
        // SSR support
        // * get value once when init
        this._WARNING_IGNORE =
          window?.localStorage?.getItem(JUNO_WARNING_IGNORE_KEY) === 'true';

        // * also provide user to set variable from window
        (window as any).__JUNO__ = this;
      }
    }

    set WARNING_IGNORE(value: boolean) {
      if (value) {
        console.log('---------------- [Juno] ----------------');
        console.log(
          `Don't recommended you close it, we should avoid use any deprecated props or methods.`,
        );
        console.log('---------------- [Juno] ----------------');
      }

      this._WARNING_IGNORE = !!value;

      window?.localStorage?.setItem(
        JUNO_WARNING_IGNORE_KEY,
        `${this._WARNING_IGNORE}`,
      );
    }

    get WARNING_IGNORE() {
      return this._WARNING_IGNORE;
    }

    set WARNING_FUNCTION(func: WarningFunctionType) {
      if (typeof func === 'function') {
        this._WARNING_FUNCTION = func;
      } else {
        console.warn('[Juno] config warningFunction is not a function.');
      }
    }

    get WARNING_FUNCTION() {
      return this._WARNING_FUNCTION;
    }
  }

  rcConfiguration = new RcConfig();
}

type RcConfigure = {
  /** is show warning message in development */
  warning?: boolean;
  /** custom warning callback */
  warningFunction?: WarningFunctionType;
};

/**
 * set config with Juno
 *
 * ## WARNING method only work in `development mode`
 */
export const configure = (options: RcConfigure) => {
  if (process.env.NODE_ENV !== 'production') {
    const { warning, warningFunction } = options;
    if (warning !== undefined) {
      rcConfiguration.WARNING_IGNORE = !warning;
    }
    if (warningFunction !== undefined) {
      rcConfiguration.WARNING_FUNCTION = warningFunction;
    }
  }
};

export { rcConfiguration };
export type { WarningFunctionType };
