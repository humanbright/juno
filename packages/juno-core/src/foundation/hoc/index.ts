export * from './withDelay';
export * from './withResponsive';
