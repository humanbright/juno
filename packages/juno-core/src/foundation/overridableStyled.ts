import {
  OverridableComponent,
  OverridableTypeMap,
} from '@material-ui/core/OverridableComponent';
import { ComponentType, FunctionComponent } from 'react';

import styled, { Interpolation, ThemedStyledProps } from './styled-components';
import { RcTheme } from './theme/theme.type';

export type CustomStyledComponentResult<T> = Interpolation<
  ThemedStyledProps<T, RcTheme>
> &
  FunctionComponent<T>;

export interface OverridableStyled {
  <M extends Omit<OverridableTypeMap, 'classKey'>>(component: ComponentType): (
    strings: TemplateStringsArray,
    ...interpolations: Array<
      Interpolation<ThemedStyledProps<M['props'], RcTheme>>
    >
  ) => OverridableComponent<M & { classKey: '' }> &
    CustomStyledComponentResult<M['props']>;
}

/**
 * just type to make styled-component work with Mui dynamic component type,
 *
 * you need custom Map type like below, define the defaultComponent type and provide that static Props
  *
 *
 * @example
 *
 * ```ts
 *  interface RcPaginationItemTypeMap<D extends React.ElementType = 'div'> {
      props: RcPaginationItemProps;
      defaultComponent: D;
    }
   ```
 */
export const overridableStyled: OverridableStyled = styled as any;
