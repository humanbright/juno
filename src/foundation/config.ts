import { ShowDeprecatedArgs } from './utils';

/* eslint-disable no-console */
const JUNO_WARNING_IGNORE_KEY = '__JUNO_WARNING_IGNORE__';

type WarningFunctionType = (
  /** warning message */
  message: string,
  /** addition options with that warning */
  options?: Pick<ShowDeprecatedArgs, 'level'>,
) => void;

class RcConfig {
  private _WARNING_IGNORE: boolean = false;

  private _WARNING_FUNCTION: WarningFunctionType = (
    message: string,
    options,
  ) => {
    console[options?.level || 'warn'](message);
  };

  constructor() {
    if (typeof window !== 'undefined') {
      // SSR support
      // * get value once when init
      this._WARNING_IGNORE =
        window?.localStorage?.getItem(JUNO_WARNING_IGNORE_KEY) === 'true';

      // * also provide user to set variable from window
      (window as any).__JUNO__ = this;
    }
  }

  set WARNING_IGNORE(value: boolean) {
    if (value) {
      console.log('---------------- [Juno] ----------------');
      console.log(
        `Don't recommended you close it, we should avoid use any deprecated props or methods.`,
      );
      console.log('---------------- [Juno] ----------------');
    }

    this._WARNING_IGNORE = !!value;

    window?.localStorage?.setItem(
      JUNO_WARNING_IGNORE_KEY,
      `${this._WARNING_IGNORE}`,
    );
  }

  get WARNING_IGNORE() {
    return this._WARNING_IGNORE;
  }

  set WARNING_FUNCTION(func: WarningFunctionType) {
    if (typeof func === 'function') {
      this._WARNING_FUNCTION = func;
    } else {
      console.warn('[Juno] config warningFunction is not a function.');
    }
  }

  get WARNING_FUNCTION() {
    return this._WARNING_FUNCTION;
  }
}

/** config instance */
const rcConfiguration = new RcConfig();

type RcConfigure = {
  /** is show warning message in development */
  warning?: boolean;
  /** custom warning callback */
  warningFunction?: WarningFunctionType;
};

/**
 * set config with Juno
 */
export const configure = ({ warning, warningFunction }: RcConfigure) => {
  if (warning !== undefined) {
    rcConfiguration.WARNING_IGNORE = !warning;
  }
  if (warningFunction !== undefined) {
    rcConfiguration.WARNING_FUNCTION = warningFunction;
  }
};

export { rcConfiguration };
export type { WarningFunctionType };
