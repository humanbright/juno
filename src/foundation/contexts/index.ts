export * from './PortalWindowContext';
export * from './ResponsiveContext';
