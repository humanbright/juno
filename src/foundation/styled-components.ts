import { useTheme } from '@material-ui/core/styles';
/* eslint-disable import/no-duplicates */
import * as styledComponents from 'styled-components';

import {
  ThemedStyledComponentsModule,
  ThemeProps as StyledThemeProps,
  InterpolationValue,
  Keyframes,
  Interpolation,
  FlattenInterpolation,
  StyleSheetManager,
  ThemedStyledProps,
} from 'styled-components';

import { RcTheme } from './theme/theme.type';

type ThemeProps = StyledThemeProps<RcTheme>;

const {
  default: styled,
  css,
  keyframes,
  createGlobalStyle,
  withTheme,
  ThemeProvider,
  ThemeConsumer,
} = (styledComponents as any) as ThemedStyledComponentsModule<RcTheme>;

type Dependencies = {
  dependencies?: (React.ComponentType | ((props: any) => JSX.Element))[];
};

const RcUseTheme = <T = RcTheme>() => useTheme<T>();

export {
  css,
  createGlobalStyle,
  keyframes,
  ThemeProvider,
  withTheme,
  ThemeConsumer,
  ThemeProps,
  Dependencies,
  InterpolationValue,
  Keyframes,
  FlattenInterpolation,
  Interpolation,
  RcUseTheme as useTheme,
  ThemedStyledProps,
  StyleSheetManager,
};

export default styled;
