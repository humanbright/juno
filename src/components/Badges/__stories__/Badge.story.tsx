import React from 'react';

import { RcBadge } from '..';
import { styled } from '../../../foundation';

export default {
  title: 'ðŸ–¤ deprecated Components/Badges',
};

const Wrapper = styled.div`
  margin-left: 20px;
`;

export const SimpleBadge = () => (
  <div>
    <div>
      <RcBadge badgeContent={1} color="primary">
        <button>primary</button>
      </RcBadge>
    </div>
    <br />
    <div>
      <RcBadge badgeContent={10} color="secondary">
        <button>secondary</button>
      </RcBadge>
    </div>
  </div>
);

export const Placement = () => (
  <Wrapper>
    <div>
      <RcBadge badgeContent={1} color="primary" placement="top-left">
        <button>top-left</button>
      </RcBadge>
      <RcBadge badgeContent={1} color="primary">
        <button>top-right</button>
      </RcBadge>
    </div>
    <div>
      <RcBadge badgeContent={1} color="primary" placement="bottom-left">
        <button>bottom-left</button>
      </RcBadge>
      <RcBadge badgeContent={1} color="primary" placement="bottom-right">
        <button>bottom-right</button>
      </RcBadge>
    </div>
  </Wrapper>
);

export const CustomizedBadge = () => (
  <div>
    <RcBadge badgeContent={<input type="checkbox" />}>
      <button>I have special badge</button>
    </RcBadge>
  </div>
);
