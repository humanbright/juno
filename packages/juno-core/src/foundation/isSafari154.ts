import { createGlobalStyle } from 'styled-components';

import { RcClasses, RcClassesProps } from './utils/classes';
import { combineProps } from './utils/combineProps';

/**
 * Conditionally apply a workaround for the CSS transition bug in Safari 15.4.
 * Remove this workaround once the Safari bug is fixed.
 */
export const isSafari154 =
  typeof navigator !== 'undefined' &&
  (/^((?!chrome|android).)*safari/i.test(navigator.userAgent) ||
    /(iPhone|iPad)/.test(navigator.userAgent)) &&
  /15\.[4-9]/i.test(navigator.userAgent);

const FixSafari154Classes = RcClasses<RcClassesProps<'paper'>>(
  ['paper'],
  'RcFixSafari154',
);

/**
 * Conditionally apply a workaround for the CSS transition bug in Safari 15.4.
 * Remove this workaround once the Safari bug is fixed.
 */
export const getSafari154Theme = (theme: any, enable?: boolean) => {
  if (enable ?? isSafari154) {
    return combineProps(
      {
        props: {
          // * set all grow related to only have one transition to make animation not shake after safari 15.4
          MuiMenu: {
            PopoverClasses: { paper: FixSafari154Classes.paper },
          },
          RcVirtualizedMenu: {
            PopoverClasses: { paper: FixSafari154Classes.paper },
          },
          MuiPopover: {
            PaperProps: { className: FixSafari154Classes.paper },
          },
          // mui tooltip prop will cover all existing props
          RcTooltip: {
            classes: {
              tooltip: FixSafari154Classes.paper,
            },
          },
          // mui grow not accept global prop
          RcGrow: {
            className: FixSafari154Classes.paper,
          },
        },
      },
      theme,
    );
  }

  return theme;
};

export const fixSafariTransitionStyle =
  'all 159ms cubic-bezier(0.4, 0, 0.2, 1) 0ms';
/**
 * A global style that can be used to apply a workaround for the CSS transition bug in Safari 15.4.
 */
export const GlobalFixSafariStyle = createGlobalStyle`
  .${FixSafari154Classes.paper} {
    transition: ${fixSafariTransitionStyle} !important;
  }
`;
