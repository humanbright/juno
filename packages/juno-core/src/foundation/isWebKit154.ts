import { createGlobalStyle } from 'styled-components';

import { RcClasses, RcClassesProps } from './utils/classes';
import { combineProps } from './utils/combineProps';

// * this whole file should be remove after webkit154 issue be fixed

/**
 * Conditionally apply a workaround for the CSS transition bug in Safari 15.4 / WebKit browsers.
 * Remove this workaround once the WebKit bug fix is released.
 *
 * follow up:
 * https://github.com/mui/material-ui/pull/31975/files
 * https://github.com/mui/material-ui/pull/32202/files
 */
export const isWebKit154 =
  typeof navigator !== 'undefined' &&
  /^((?!chrome|android).)*(safari|mobile)/i.test(navigator.userAgent) &&
  /(os |version\/)15(.|_)[4-9]/i.test(navigator.userAgent);

const FixWebKit154Classes = RcClasses<RcClassesProps<'paper'>>(
  ['paper'],
  'RcFixWebKit154',
);

/**
 * Conditionally apply a workaround for the CSS transition bug in Safari 15.4.
 * Remove this workaround once the Safari bug is fixed.
 */
export const getWebKit154Theme = (theme: any, enable?: boolean) => {
  if (enable ?? isWebKit154) {
    return combineProps(
      {
        props: {
          // * set all grow related to only have one transition to make animation not shake after safari 15.4
          MuiMenu: {
            PopoverClasses: { paper: FixWebKit154Classes.paper },
          },
          RcVirtualizedMenu: {
            PopoverClasses: { paper: FixWebKit154Classes.paper },
          },
          MuiPopover: {
            PaperProps: { className: FixWebKit154Classes.paper },
          },
          // mui tooltip prop will cover all existing props
          RcTooltip: {
            classes: {
              tooltip: FixWebKit154Classes.paper,
            },
          },
          // mui grow not accept global prop
          RcGrow: {
            className: FixWebKit154Classes.paper,
          },
        },
      },
      theme,
    );
  }

  return theme;
};

export const fixWebKitTransitionStyle =
  'all 159ms cubic-bezier(0.4, 0, 0.2, 1) 0ms';

/**
 * A global style that can be used to
 * apply a workaround for the CSS transition bug after WebKit 15.4.
 */
export const GlobalFixWebKitStyle = createGlobalStyle`
  .${FixWebKit154Classes.paper} {
    transition: ${fixWebKitTransitionStyle} !important;
  }
`;
