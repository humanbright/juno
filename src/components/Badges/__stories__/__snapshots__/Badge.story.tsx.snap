// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots ðŸ–¤ deprecated Components/Badges Customized Badge 1`] = `
.c0 .badge {
  width: 24px;
  height: 24px;
  top: -12px;
  right: -12px;
  bottom: auto;
  left: auto;
}

<div>
  <span
    className="MuiBadge-root c0"
    placement="top-right"
  >
    <button>
      I have special badge
    </button>
    <span
      className="MuiBadge-badge badge MuiBadge-anchorOriginTopRightRectangle"
    >
      <input
        type="checkbox"
      />
    </span>
  </span>
</div>
`;

exports[`Storyshots ðŸ–¤ deprecated Components/Badges Placement 1`] = `
.c1 .badge {
  width: 24px;
  height: 24px;
  top: -12px;
  right: auto;
  bottom: auto;
  left: -12px;
}

.c2 .badge {
  width: 24px;
  height: 24px;
  top: -12px;
  right: -12px;
  bottom: auto;
  left: auto;
}

.c3 .badge {
  width: 24px;
  height: 24px;
  top: auto;
  right: auto;
  bottom: -12px;
  left: -12px;
}

.c4 .badge {
  width: 24px;
  height: 24px;
  top: auto;
  right: -12px;
  bottom: -12px;
  left: auto;
}

.c0 {
  margin-left: 20px;
}

<div
  className="c0"
>
  <div>
    <span
      className="MuiBadge-root c1"
      placement="top-left"
    >
      <button>
        top-left
      </button>
      <span
        className="MuiBadge-badge badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary"
      >
        1
      </span>
    </span>
    <span
      className="MuiBadge-root c2"
      placement="top-right"
    >
      <button>
        top-right
      </button>
      <span
        className="MuiBadge-badge badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary"
      >
        1
      </span>
    </span>
  </div>
  <div>
    <span
      className="MuiBadge-root c3"
      placement="bottom-left"
    >
      <button>
        bottom-left
      </button>
      <span
        className="MuiBadge-badge badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary"
      >
        1
      </span>
    </span>
    <span
      className="MuiBadge-root c4"
      placement="bottom-right"
    >
      <button>
        bottom-right
      </button>
      <span
        className="MuiBadge-badge badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary"
      >
        1
      </span>
    </span>
  </div>
</div>
`;

exports[`Storyshots ðŸ–¤ deprecated Components/Badges Simple Badge 1`] = `
.c0 .badge {
  width: 24px;
  height: 24px;
  top: -12px;
  right: -12px;
  bottom: auto;
  left: auto;
}

<div>
  <div>
    <span
      className="MuiBadge-root c0"
      placement="top-right"
    >
      <button>
        primary
      </button>
      <span
        className="MuiBadge-badge badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary"
      >
        1
      </span>
    </span>
  </div>
  <br />
  <div>
    <span
      className="MuiBadge-root c0"
      placement="top-right"
    >
      <button>
        secondary
      </button>
      <span
        className="MuiBadge-badge badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorSecondary"
      >
        10
      </span>
    </span>
  </div>
</div>
`;
